sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/HashUtils.scala

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf
+package crypto
+
+import com.daml.scalautil.Statement.discard
+import com.digitalasset.daml.lf.crypto.Hash.underlyingHashLength
+import com.digitalasset.daml.lf.data.Bytes
+
+import java.io.OutputStream
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+
+object HashUtils {
+
+  private val lineSeparator = System.getProperty("line.separator")
+  private[crypto] def formatByteToHexString(byte: Byte): String = String.format("%02X", byte)
+
+  /** Extension of OutputStream with additional methods to help exporting the encoding with context information
+    */
+  abstract class ContextAwareOutputStream extends OutputStream {
+    def write(byte: Byte): Unit
+    def writeContext(context: String): Unit
+  }
+
+  /** Implementation of ContextAwareOutputStream for debugging.
+    * Support contextual info and writes the encoding to a String
+    */
+  class DebugStringOutputStream extends ContextAwareOutputStream {
+    private val sb = new StringBuilder()
+    def result: String = sb.result()
+    override def write(b: Int): Unit =
+      discard(sb.append(Bytes.fromByteArray(Array(b.toByte)).toHexString))
+    override def write(b: Array[Byte], off: Int, len: Int) = {
+      discard(sb.append(Bytes.fromByteArray(b.slice(off, off + len)).toHexString))
+    }
+    override def write(byte: Byte): Unit =
+      discard(sb.append(formatByteToHexString(byte)))
+
+    def writeContext(context: String): Unit = discard(sb.append(context))
+
+    override def close(): Unit = sb.clear()
+  }
+
+  /** Interface for message digest. Allows to swap out default implementation for debug implementation.
+    */
+  private[crypto] trait MessageDigestWithContext {
+    def messageDigest: MessageDigest
+    def update(a: Byte, context: Option[String]): Unit
+    def update(a: ByteBuffer, context: Option[String]): Unit
+    def update(a: Array[Byte], context: Option[String]): Unit
+    def digest(buf: Array[Byte], offset: Int, len: Int): Int
+    def addContext(context: String): Unit
+    def isContextAware: Boolean
+  }
+
+  /** Default Message Digest implementation. Ignores context and simply feeds all bytes to the delegate MessageDigest.
+    * @param messageDigest MessageDigest delegate which calculates the hash.
+    */
+  private[crypto] class DefaultMessageDigest(override val messageDigest: MessageDigest)
+      extends MessageDigestWithContext {
+    override val isContextAware: Boolean = false
```

Comment:
There's not much point in storing a boolean constant as a field of the objects of a class.
```suggestion
    override def isContextAware: Boolean = false
```